message/validation/common_checks.go

```diff
@@ -13,70 +14,108 @@ func (mv *messageValidator) committeeRole(role spectypes.RunnerRole) bool {
-		e := ErrLateMessage
+	if lateness := mv.messageLateness(messageSlot, role, receivedAt); lateness > clockErrorTolerance {
+		e := ErrLateSlotMessage
 		e.got = fmt.Sprintf("late by %v", lateness)
 		return e
 	}
 
 	return nil
 }
 
-func (mv *messageValidator) earlyMessage(slot phase0.Slot, receivedAt time.Time) bool {
-	return mv.netCfg.Beacon.GetSlotEndTime(mv.netCfg.Beacon.EstimatedSlotAtTime(receivedAt.Unix())).
-		Add(-clockErrorTolerance).Before(mv.netCfg.Beacon.GetSlotStartTime(slot))
+// messageEarliness returns how early message is or 0 if it's not
+func (mv *messageValidator) messageEarliness(slot phase0.Slot, receivedAt time.Time) time.Duration {
+	return mv.netCfg.Beacon.GetSlotStartTime(slot).Sub(receivedAt)
 }
 
-func (mv *messageValidator) lateMessage(slot phase0.Slot, role spectypes.RunnerRole, receivedAt time.Time) time.Duration {
+// messageLateness returns how late message is or 0 if it's not
+func (mv *messageValidator) messageLateness(slot phase0.Slot, role spectypes.RunnerRole, receivedAt time.Time) time.Duration {
 	var ttl phase0.Slot
 	switch role {
 	case spectypes.RoleProposer, spectypes.RoleSyncCommitteeContribution:
 		ttl = 1 + lateSlotAllowance
 	case spectypes.RoleCommittee, spectypes.RoleAggregator:
-		ttl = 32 + lateSlotAllowance
+		ttl = phase0.Slot(mv.netCfg.Beacon.SlotsPerEpoch()) + lateSlotAllowance
 	case spectypes.RoleValidatorRegistration, spectypes.RoleVoluntaryExit:
 		return 0
 	}
 
 	deadline := mv.netCfg.Beacon.GetSlotStartTime(slot + ttl).
-		Add(lateMessageMargin).Add(clockErrorTolerance)
+		Add(lateMessageMargin)
 
-	return mv.netCfg.Beacon.GetSlotStartTime(mv.netCfg.Beacon.EstimatedSlotAtTime(receivedAt.Unix())).
-		Sub(deadline)
+	return receivedAt.Sub(deadline)
 }
 
 func (mv *messageValidator) validateDutyCount(
-	validatorIndices []phase0.ValidatorIndex,
-	state *SignerState,
 	msgID spectypes.MessageID,
-	newDutyInSameEpoch bool,
+	msgSlot phase0.Slot,
+	validatorIndices []phase0.ValidatorIndex,
+	signerStateBySlot *treemap.Map,
 ) error {
-	var dutyLimit int
+	msgEpoch := mv.netCfg.Beacon.EstimatedEpochAtSlot(msgSlot)
+	dutyCount := 0
+	signerStateBySlot.Each(func(slot any, state any) {
+		if mv.netCfg.Beacon.EstimatedEpochAtSlot(slot.(phase0.Slot)) == msgEpoch {
+			dutyCount++
+		}
+	})
+	if _, ok := signerStateBySlot.Get(msgSlot); !ok {
+		dutyCount++
+	}
 
+	dutyLimit, exists := mv.dutyLimit(msgID, validatorIndices)
+	if !exists {
+		return nil
+	}
+
+	if dutyCount > dutyLimit {
+		err := ErrTooManyDutiesPerEpoch
+		err.got = fmt.Sprintf("%v (role %v)", dutyCount, msgID.GetRoleType())
+		err.want = fmt.Sprintf("less than %v", dutyLimit)
+		return err
+	}
+
+	return nil
+}
+
+func (mv *messageValidator) dutyLimit(msgID spectypes.MessageID, validatorIndices []phase0.ValidatorIndex) (int, bool) {
 	switch msgID.GetRoleType() {
 	case spectypes.RoleAggregator, spectypes.RoleValidatorRegistration, spectypes.RoleVoluntaryExit:
-		dutyLimit = 2
+		return 2, true
```

Comment:
VoluntaryExit is risky to put here because we don't limit the rate at which nodes can start this duty, its dictated purely by contract events.

We should probably hook up to the event syncer in some way to ask if such duty exists (if it received an event for this validator and this slot). 

Regarding ValidatorRegistration, its basically scheduled once per 10 epochs for each validator, however it might be necessary to allow 2 per epoch, since we're also performing it immediately after validators register, so this helps avoid ignoring/rejecting cases where registration and scheduling happened to fall on the same epoch for some validator. With that said, we should likely come up with a better solution long term.

Reply:
@moshe-blox I agree, this sounds good to me but it looks like something worth a separate PR because this PR doesn't change this logic

Reply:
@nkryuchkov +1
thanks

Reply:
@moshe-blox and I decided that VoluntaryExit changes won't be implemented now, so only ValidatorRegistration will be done